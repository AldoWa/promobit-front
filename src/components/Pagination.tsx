import { Button, HStack } from '@chakra-ui/react';
import React, { useCallback, useEffect, useState } from 'react';

interface PaginationProps {
  handlePage: (page: number) => void;
  maxTotalPages: number;
  actualPage: number;
}

const Pagination: React.FC<PaginationProps> = ({ handlePage, maxTotalPages, actualPage }) => {
  const [pagination, setPagination] = useState<number[]>([1, 2, 3, 4, 5])


  const handleNewPage = async (page: number) => {
    await handlePage(page)
    createPagination(page)
  }

  const createPagination = useCallback((page: number) => {
    if (page <= 2) {
      setPagination([1, 2, 3, 4, 5])
      return
    }

    if (page === maxTotalPages) {
      setPagination([page - 4, page - 3, page - 2, page - 1, page])
      return
    }

    if (page === maxTotalPages - 1) {
      setPagination([page - 3, page - 2, page - 1, page, page + 1])
      return
    }
    setPagination([page - 2, page - 1, page, page + 1, page + 2])
  }, [maxTotalPages])

  useEffect(() => {
    createPagination(actualPage)
  }, [actualPage, createPagination])

  return (
    <HStack justifyContent='center' mt={70} data-testid="pagination-stack">
      {actualPage > 1 && <Button onClick={() => handleNewPage(1)} data-testid="pagination-first-page">Primeira</Button>}
      {actualPage > 1 && <Button onClick={() => handleNewPage(actualPage - 1)}>{'<'}</Button>}
      <HStack data-testid="pagination-pages">
        {pagination.map(page => (
          <Button
            key={page}
            onClick={() => handleNewPage(page)}
            color={actualPage === page ? 'white' : 'black'}
            bg={actualPage === page ? 'orange.500' : 'gray.100'}
            _hover={{ bg: 'orange.500', color: 'white' }}
            aria-labelledby='Ir para a página'
          >{page}</Button>
        ))}
      </HStack>
      {actualPage < maxTotalPages && (
        <>
          <Button onClick={() => handleNewPage(actualPage + 1)}>{'>'}</Button>
          <Button onClick={() => handleNewPage(maxTotalPages)} data-testid="pagination-last-page">Última</Button>
        </>
      )}
    </HStack>
  );
};

export default Pagination;
