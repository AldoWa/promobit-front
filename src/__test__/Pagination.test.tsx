import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import Pagination from "../components/Pagination";

describe('Pagination', () => {
  it('Should render the Pagination', () => {
    render(<Pagination
      actualPage={1}
      maxTotalPages={10}
      handlePage={() => {}}
     
    />)

    const paginationStack = screen.getByTestId('pagination-stack');
    const pages = screen.getByTestId('pagination-pages').childNodes;

    expect(paginationStack).toBeInTheDocument();
    expect(pages.length).toBe(5);
    expect(pages[0]).toHaveTextContent('1');
  });

  it('Should render the item to go to the last page and not render the first', () => {
    render(<Pagination
      actualPage={1}
      maxTotalPages={10}
      handlePage={() => {}}
    />)

    const paginationLastPage = screen.queryByTestId('pagination-last-page');
    expect(paginationLastPage).toBeInTheDocument();
    
    const paginationFirstPage = screen.queryByTestId('pagination-first-page');
    expect(paginationFirstPage).not.toBeInTheDocument();
  });

  it('Should render the item to go to the first page and not render the last', () => {
    render(<Pagination
      actualPage={10}
      maxTotalPages={10}
      handlePage={() => {}}
    />)

    const paginationLastPage = screen.queryByTestId('pagination-last-page');
    expect(paginationLastPage).not.toBeInTheDocument();
    
    const paginationFirstPage = screen.queryByTestId('pagination-first-page');
    expect(paginationFirstPage).toBeInTheDocument();
  });

  it('Should render the pagination items correct', () => {
    render(<Pagination
      actualPage={5}
      maxTotalPages={10}
      handlePage={() => {}}
    />)

    const pages = screen.getByTestId('pagination-pages').childNodes;

    expect(pages[0]).toHaveTextContent('3');
    expect(pages[1]).toHaveTextContent('4');
    expect(pages[2]).toHaveTextContent('5');
    expect(pages[3]).toHaveTextContent('6');
    expect(pages[4]).toHaveTextContent('7');
  
  })

  it('Should call handlePage when click on a page', async () => {
    const handlePage = jest.fn();
    
    render(<Pagination
      actualPage={5}
      maxTotalPages={10}
      handlePage={handlePage}
    />)
   
    const pages = screen.getByTestId('pagination-pages').childNodes;

    fireEvent.click(pages[0]);
    await waitFor(() => {
      expect(handlePage).toHaveBeenCalledTimes(1);
    })
  })

  it("Should match snapshot", () => {
    const { container } = render(<Pagination
      actualPage={5}
      maxTotalPages={10}
      handlePage={() => {}}
    />)
    expect(container).toMatchSnapshot();
  });
});