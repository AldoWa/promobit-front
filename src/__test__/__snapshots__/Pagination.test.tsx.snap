// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination Should match snapshot 1`] = `
<div>
  <div
    class="chakra-stack css-z38eug"
    data-testid="pagination-stack"
  >
    <button
      class="chakra-button css-4xx2wk"
      data-testid="pagination-first-page"
      type="button"
    >
      Primeira
    </button>
    <button
      class="chakra-button css-4xx2wk"
      type="button"
    >
      &lt;
    </button>
    <div
      class="chakra-stack css-1igwmid"
      data-testid="pagination-pages"
    >
      <button
        aria-labelledby="Ir para a página"
        class="chakra-button css-ljn2h4"
        type="button"
      >
        3
      </button>
      <button
        aria-labelledby="Ir para a página"
        class="chakra-button css-ljn2h4"
        type="button"
      >
        4
      </button>
      <button
        aria-labelledby="Ir para a página"
        class="chakra-button css-1etlj6"
        type="button"
      >
        5
      </button>
      <button
        aria-labelledby="Ir para a página"
        class="chakra-button css-ljn2h4"
        type="button"
      >
        6
      </button>
      <button
        aria-labelledby="Ir para a página"
        class="chakra-button css-ljn2h4"
        type="button"
      >
        7
      </button>
    </div>
    <button
      class="chakra-button css-4xx2wk"
      type="button"
    >
      &gt;
    </button>
    <button
      class="chakra-button css-4xx2wk"
      data-testid="pagination-last-page"
      type="button"
    >
      Última
    </button>
  </div>
</div>
`;
